[Tooling] Add --exclude flag to Generator to support field removal testing

djptek opened the pull request:
- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)? Y
- Have you followed the [contributor guidelines](https://github.com/elastic/ecs/blob/master/CONTRIBUTING.md)? Y
- For proposing substantial changes or additions to the schema, have you reviewed the [RFC process] (https://github.com/elastic/ecs/blob/master/rfcs/README.md)? n/a
- If submitting code/script changes, have you verified all tests pass locally using `make test`? Y
- If submitting schema/fields updates, have you generated new artifacts by running `make` and committed those changes? n/a
- Is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed. Y
- Have you added an entry to the [CHANGELOG.next.md] WIP (https://github.com/elastic/ecs/blob/master/CHANGELOG.next.md)?

djptek commented:
Note: Moved to Draft while fixing regression

djptek commented:
Regression fixed: the files in folder `.rfcs/text/0017` have been included temporarily as I used those for testing, these will not be included in the final commit

ebeahan commented on scripts/generator.py:
```diff
@@ -50,7 +50,7 @@ def main():
     cleaner.clean(fields, strict=args.strict)
     finalizer.finalize(fields)
-    fields = subset_filter.filter(fields, args.subset, out_dir)
+    fields = subset_filter.filter(fields, args.subset, args.exclude, out_dir)
```
What do you think about moving `exclude` into its own module instead of extending `subset_filter`?

Outside of the functions for loading the files, it looks like the two are independent of each other. Might help when it comes to writing tests to have things more isolated.

djptek replied:
OK, I'll refactor :+1:

djptek replied:
I've moved that to `exclude_filter`. Where functions would have been exactly duplicated, I kept these as imports from `subset_filter` to keep things DRY

djptek replied:
I did a more extensive refactor, see: https://github.com/elastic/ecs/pull/1411#issuecomment-841362269

ebeahan commented on rfcs/text/0017/exclude-set.yml:
```diff
@@ -0,0 +1,4 @@
+---
```
Looks like you have some good test case built! These will be good examples to help build out some of the unit testing.

djptek replied:
thx!

ebeahan commented on scripts/schema/subset_filter.py:
```diff
@@ -18,6 +20,45 @@ def filter(fields, subset_file_globs, out_dir):
+        if len(node_path) == 1:
+            b4 = fields.copy()
+            fields.pop(node_path[0])
+            print("removed field:", (set(b4.keys() ^ set(fields.keys()))).pop())
```
I'm finding the set operation here a bit confusing to follow. Is the intent to show the field that was removed?

`.pop()` will remove the value but also returns the element if you want to store it in a variable.

djptek replied:
Yeah, I didn't use the return value of pop because the field itself can be huge with subfields and comments and all and I didn't want to swamp the output. The set operation was useful to me while debugging but I agree it is OTT now so I'll find a more succinct way, maybe parse the pop return value after all :+1:

djptek replied:
I've simplified this to take advantage of `flat_name`, test output now looks like:

```
djptek@Elastic-MacBook-Pro ecs % python scripts/generator.py --exclude="./rfcs/text/0017/exclude-set.yml"                (exclude_set)ecs
Loading schemas from local files
Running generator. ECS version 2.0.0-dev
Removed field log.original
djptek@Elastic-MacBook-Pro ecs % python scripts/generator.py --exclude="./rfcs/text/0017/exclude-set-pass2.yml"          (exclude_set)ecs
Loading schemas from local files
Running generator. ECS version 2.0.0-dev
Removed field log.original
Removed field log.file.path
djptek@Elastic-MacBook-Pro ecs % python scripts/generator.py --exclude="./rfcs/text/0017/exclude-set-fail*.yml"          (exclude_set)ecs
Loading schemas from local files
Running generator. ECS version 2.0.0-dev
No match for exclusion: test_fail1_nonexistent_root.file.path
No match for exclusion: test_fail2_nonexistent_node.path
No match for exclusion: path.test_fail3_nonexistent_leaf
```

djptek replied:
All print statements now removed in favour of unit tests, see: https://github.com/elastic/ecs/pull/1411#issuecomment-841362269

djptek commented:
moved to draft while updating unit tests

djptek commented:
Refactor: 
- moved common yml loading code & related tests from `subset_filter` to `loader`
- unstaged simplified tests from 0017
- added unit tests for `exclude_filter`
- converted warnings to exceptions in `exclude_filter`
- removed `print` statements from `exclude filter`

Move back to "ready for review"

kgeller commented on scripts/schema/loader.py:
```diff
@@ -251,3 +251,33 @@ def merge_fields(a, b):
+
+
+def eval_globs(globs):
+    '''Accepts an array of glob patterns or file names, returns the array of actual files'''
```
Nit: should we use `"` here instead of `'` to be consistent with the `exclude_filter`?

djptek replied:
Good catch @kgeller - I checked and that was also inconsistent elsewhere so I've normalized to `"""` as per [PEP-0257](https://www.python.org/dev/peps/pep-0257/)

ebeahan commented on scripts/schema/exclude_filter.py:
```diff
@@ -0,0 +1,62 @@
+import glob
```
Do `glob`, `yaml`, or `os` modules get called directly here? If not, we can clean up these imports.

djptek replied:
nope, removed unused imports here and also `loader.py`and `subset_filter.py`

ebeahan reviewed (commented):
Thanks for refactoring a bit and adding the unit tests!

djptek commented on scripts/generator.py:
```diff
@@ -70,6 +72,8 @@ def argument_parser():
                                                        Note that "--include experimental/schemas" will also respect this git ref.')
     parser.add_argument('--include', nargs='+',
                         help='include user specified directory of custom field definitions')
+    parser.add_argument('--exclude', nargs='+',
```
Good catch, added to `USAGE.md`

djptek commented on scripts/tests/unit/test_schema_exclude_filter.py:
```diff
@@ -0,0 +1,97 @@
+from schema import exclude_filter
+import mock
+import os
+import pprint
```
removed unused imports :+1:
